lightning/src/ln/onion_utils.rs

```diff
@@ -725,7 +725,8 @@ pub(crate) fn decode_next_hop<D: DecodeInput, R: ReadableArgs<D::Arg>, N: NextPa
 				return Ok((msg, None)); // We are the final destination for this packet
 			} else {
 				let mut new_packet_bytes = N::new(hop_data.len());
-				let read_pos = chacha_stream.read(new_packet_bytes.as_mut()).unwrap();
+				let new_packet_bytes_len = hop_data.len() - chacha_stream.bytes_read();
+				chacha_stream.read_exact(&mut new_packet_bytes.as_mut()[..new_packet_bytes_len]).unwrap();
```

Comment:
Instead of keeping track of `bytes_read` in `ChaChaReader`, could you use the `Cursor::position` of the underlying `ChaChaReader::read`?

Reply:
`ChaChaReader` takes a generic `io::Read` :/ don't think we can use `Cursor::position`?

Reply:
Oh, I see what you mean. I just tried this but it looks like we'd need a `mut` borrow of the `Cursor` to create the `ChaChaReader`, and `position` then tries to take an immutable borrow and errors.

Reply:
The `mut` borrow is on `ChaCha20`, not on the `Cursor`. So this should work:

```rust
let new_packet_bytes_len = hop_data.len() - chacha_stream.read.position() as usize;
```

Reply:
Oh nice! fixed